File: app/src/main/java/com/dmi/perfectreader/control/TapZoneConfiguration.java
package com.dmi.perfectreader.control;

public enum TapZoneConfiguration {
    /**
     *  _ _ _ _ _ _
     * |   |   |   |
     * |_ _|_ _|_ _|
     * |   |   |   |
     * |_ _|_ _|_ _|
     * |   |   |   |
     * |_ _|_ _|_ _|
     */
    NINE_SQUARES,

    /**
     *  _ _ _ _ _ _
     * |     |     |
     * |     |     |
     * |_ _ _|_ _ _|
     * |     |     |
     * |     |     |
     * |_ _ _|_ _ _|
     */
    FOUR_SQUARES,

    /**
     *  _ _ _ _ _ _
     * |     |     |
     * |_ _ _|_ _ _|
     * |     |     |
     * |_ _ _|_ _ _|
     * |     |     |
     * |_ _ _|_ _ _|
     */
    THREE_SQUARES_ON_LEFT_AND_RIGHT,

    /**
     *  _ _ _ _ _ _
     * |   |   |   |
     * |   |   |   |
     * |_ _|_ _|_ _|
     * |   |   |   |
     * |   |   |   |
     * |_ _|_ _|_ _|
     */
    THREE_SQUARES_ON_TOP_AND_BOTTOM,

    /**
     *  _ _ _ _
     * |\     /|
     * | \   / |
     * |  \ /  |
     * |  / \  |
     * | /   \ |
     * |/_ _ _\|
     */
    FOUR_TRIANGLES,

    /**
     *  _ _ _ _ _
     * |\       /|
     * | \ _ _ / |
     * |  |   |  |
     * |  |_ _|  |
     * | /     \ |
     * |/_ _ _ _\|
     */
    FOUR_TRAPEZOIDS_AND_SQUARE;

    public TapZone getAt(float xPart, float yPart) {
        if (xPart < 0) xPart = 0;
        if (yPart < 0) yPart = 0;
        if (xPart > 1) xPart = 1;
        if (yPart > 1) yPart = 1;

        switch (this) {
            case NINE_SQUARES: {
                if (xPart < 1 / 3F && yPart < 1 / 3F) {
                    return TapZone.TOP_LEFT;
                } else if (xPart > 2 / 3F && yPart < 1 / 3F) {
                    return TapZone.TOP_RIGHT;
                } else if (xPart < 1 / 3F && yPart > 2 / 3F) {
                    return TapZone.BOTTOM_LEFT;
                } else if (xPart > 2 / 3F && yPart > 2 / 3F) {
                    return TapZone.BOTTOM_RIGHT;
                } else if (yPart < 1 / 3F) {
                    return TapZone.TOP;
                } else if (yPart > 2 / 3F) {
                    return TapZone.BOTTOM;
                } else if (xPart < 1 / 3F) {
                    return TapZone.LEFT;
                } else if (xPart > 2 / 3F) {
                    return TapZone.RIGHT;
                } else {
                    return TapZone.CENTER;
                }
            }

            case FOUR_SQUARES: {
                if (xPart < 1 / 2F && yPart < 1 / 2F) {
                    return TapZone.TOP_LEFT;
                } else if (xPart >= 1 / 2F && yPart < 1 / 2F) {
                    return TapZone.TOP_RIGHT;
                } else if (xPart < 1 / 2F && yPart >= 1 / 2F) {
                    return TapZone.BOTTOM_LEFT;
                } else {
                    return TapZone.BOTTOM_RIGHT;
                }
            }

            case THREE_SQUARES_ON_LEFT_AND_RIGHT: {
                if (xPart < 1 / 2F && yPart < 1 / 3F) {
                    return TapZone.TOP_LEFT;
                } else if (xPart >= 1 / 2F && yPart < 1 / 3F) {
                    return TapZone.TOP_RIGHT;
                } else if (xPart < 1 / 2F && yPart > 2 / 3F) {
                    return TapZone.BOTTOM_LEFT;
                } else if (xPart >= 1 / 2F && yPart > 2 / 3F) {
                    return TapZone.BOTTOM_RIGHT;
                } else if (xPart < 1 / 2F) {
                    return TapZone.LEFT;
                } else {
                    return TapZone.RIGHT;
                }
            }

            case THREE_SQUARES_ON_TOP_AND_BOTTOM: {
                if (xPart < 1 / 3F && yPart < 1 / 2F) {
                    return TapZone.TOP_LEFT;
                } else if (xPart > 2 / 3F && yPart < 1 / 2F) {
                    return TapZone.TOP_RIGHT;
                } else if (xPart < 1 / 3F && yPart >= 1 / 2F) {
                    return TapZone.BOTTOM_LEFT;
                } else if (xPart > 2 / 3F && yPart >= 1 / 2F) {
                    return TapZone.BOTTOM_RIGHT;
                } else if (yPart < 1 / 2F) {
                    return TapZone.TOP;
                } else {
                    return TapZone.BOTTOM;
                }
            }

            case FOUR_TRIANGLES: {
                boolean bottomOrLeft = xPart < yPart;
                boolean topOrLeft = xPart + yPart < 1;
                if (bottomOrLeft && topOrLeft) {
                    return TapZone.LEFT;
                } else if (!bottomOrLeft && topOrLeft) {
                    return TapZone.TOP;
                } else if (bottomOrLeft) {
                    return TapZone.RIGHT;
                } else {
                    return TapZone.BOTTOM;
                }
            }

            case FOUR_TRAPEZOIDS_AND_SQUARE: {
                boolean inSquare = xPart >= 1 / 3F && xPart <= 2 / 3F && yPart >= 1 / 3F && yPart <= 2 / 3F;
                boolean bottomOrLeft = xPart < yPart;
                boolean topOrLeft = xPart + yPart < 1;
                if (inSquare) {
                    return TapZone.CENTER;
                } else if (bottomOrLeft && topOrLeft) {
                    return TapZone.LEFT;
                } else if (!bottomOrLeft && topOrLeft) {
                    return TapZone.TOP;
                } else if (bottomOrLeft) {
                    return TapZone.RIGHT;
                } else {
                    return TapZone.BOTTOM;
                }
            }

            default:
                throw new IllegalStateException();
        }
    }
}

File: app/src/main/kotlin/com/dmi/perfectreader/fragment/bookcontrol/entity/TapZoneConfiguration.kt
package com.dmi.perfectreader.fragment.bookcontrol.entity

import com.dmi.util.lang.clamp

enum class TapZoneConfiguration {
    /**
     *  _ _ _ _ _ _
     * |   |   |   |
     * |_ _|_ _|_ _|
     * |   |   |   |
     * |_ _|_ _|_ _|
     * |   |   |   |
     * |_ _|_ _|_ _|
     */
    NINE,

    /**
     *  _ _ _ _ _ _
     * |           |
     * |           |
     * |           |
     * |           |
     * |           |
     * |_ _ _ _ _ _|
     */
    SINGLE,

    /**
     *  _ _ _ _ _ _
     * |     |     |
     * |     |     |
     * |_ _ _|_ _ _|
     * |     |     |
     * |     |     |
     * |_ _ _|_ _ _|
     */
    FOUR,

    /**
     *  _ _ _ _ _ _
     * |     |     |
     * |_ _ _|_ _ _|
     * |     |     |
     * |_ _ _|_ _ _|
     * |     |     |
     * |_ _ _|_ _ _|
     */
    TWO_TOP_BOTTOM_CENTER,

    /**
     *  _ _ _ _ _ _
     * |   |   |   |
     * |_ _|_ _|_ _|
     * |     |     |
     * |_ _ _|_ _ _|
     * |   |   |   |
     * |_ _|_ _|_ _|
     */
    THREE_TOP_BOTTOM_TWO_CENTER,

    /**
     *  _ _ _ _ _ _
     * |     |     |
     * |_ _ _|_ _ _|
     * |   |   |   |
     * |_ _|_ _|_ _|
     * |     |     |
     * |_ _ _|_ _ _|
     */
    TWO_TOP_BOTTOM_THREE_CENTER,

    /**
     *  _ _ _ _ _ _
     * |   |   |   |
     * |   |   |   |
     * |_ _|_ _|_ _|
     * |   |   |   |
     * |   |   |   |
     * |_ _|_ _|_ _|
     */
    TWO_LEFT_RIGHT_CENTER,

    /**
     *  _ _ _ _ _ _
     * |   |   |   |
     * |_ _|   |_ _|
     * |   |_ _|   |
     * |_ _|   |_ _|
     * |   |   |   |
     * |_ _|_ _|_ _|
     */
    THREE_LEFT_RIGHT_TWO_CENTER,

    /**
     *  _ _ _ _ _ _
     * |   |   |   |
     * |   |_ _|   |
     * |_ _|   |_ _|
     * |   |_ _|   |
     * |   |   |   |
     * |_ _|_ _|_ _|
     */
    TWO_LEFT_RIGHT_THREE_CENTER,

    /**
     *  _ _ _ _ _ _
     * | /       \ |
     * |/ \ _ _ / \|
     * |   |   |   |
     * |   |_ _|   |
     * |\ /     \ /|
     * |_\_ _ _ _/_|
     */
    TRIANGLE_SIDE_CENTER_CORNERS,

    /**
     *  _ _ _ _
     * |\     /|
     * | \   / |
     * |  \ /  |
     * |  / \  |
     * | /   \ |
     * |/_ _ _\|
     */
    TRIANGLE_SIDES,

    /**
     *  _ _ _ _ _
     * |\       /|
     * | \ _ _ / |
     * |  |   |  |
     * |  |_ _|  |
     * | /     \ |
     * |/_ _ _ _\|
     */
    TRIANGLE_SIDES_CENTER,

    /**
     *  _ _ _ _ _
     * | /     \ |
     * |/ \   / \|
     * |   \ /   |
     * |   / \   |
     * |\ /   \ /|
     * |_\_ _ _/_|
     */
    TRIANGLE_SIDE_CORNERS;

    fun getAt(xPart: Float, yPart: Float): TapZone {
        val x = clamp(xPart, 0F, 1F)
        val y = clamp(yPart, 0F, 1F)

        return when (this) {
            NINE -> when {
                x < 1 / 3f && y < 1 / 3f -> TapZone.TOP_LEFT
                x > 2 / 3f && y < 1 / 3f -> TapZone.TOP_RIGHT
                x < 1 / 3f && y > 2 / 3f -> TapZone.BOTTOM_LEFT
                x > 2 / 3f && y > 2 / 3f -> TapZone.BOTTOM_RIGHT
                y < 1 / 3f -> TapZone.TOP
                y > 2 / 3f -> TapZone.BOTTOM
                x < 1 / 3f -> TapZone.LEFT
                x > 2 / 3f -> TapZone.RIGHT
                else -> TapZone.CENTER
            }

            SINGLE -> TapZone.CENTER

            FOUR -> when {
                x < 1 / 2f && y < 1 / 2f -> TapZone.TOP_LEFT
                x >= 1 / 2f && y < 1 / 2f -> TapZone.TOP_RIGHT
                x < 1 / 2f && y >= 1 / 2f -> TapZone.BOTTOM_LEFT
                else -> TapZone.BOTTOM_RIGHT
            }

            TWO_TOP_BOTTOM_CENTER -> when {
                x < 1 / 2f && y < 1 / 3f -> TapZone.TOP_LEFT
                x >= 1 / 2f && y < 1 / 3f -> TapZone.TOP_RIGHT
                x < 1 / 2f && y > 2 / 3f -> TapZone.BOTTOM_LEFT
                x >= 1 / 2f && y > 2 / 3f -> TapZone.BOTTOM_RIGHT
                x < 1 / 2f -> TapZone.LEFT
                else -> TapZone.RIGHT
            }

            THREE_TOP_BOTTOM_TWO_CENTER -> when {
                x < 1 / 3f && y < 1 / 3f -> TapZone.TOP_LEFT
                x > 2 / 3f && y < 1 / 3f -> TapZone.TOP_RIGHT
                x < 1 / 3f && y > 2 / 3f -> TapZone.BOTTOM_LEFT
                x > 2 / 3f && y > 2 / 3f -> TapZone.BOTTOM_RIGHT
                y < 1 / 3f -> TapZone.TOP
                y > 2 / 3f -> TapZone.BOTTOM
                x < 1 / 2f -> TapZone.LEFT
                else -> TapZone.RIGHT
            }

            TWO_TOP_BOTTOM_THREE_CENTER -> when {
                x < 1 / 2f && y < 1 / 3f -> TapZone.TOP_LEFT
                x > 1 / 2f && y < 1 / 3f -> TapZone.TOP_RIGHT
                x < 1 / 2f && y > 2 / 3f -> TapZone.BOTTOM_LEFT
                x > 1 / 2f && y > 2 / 3f -> TapZone.BOTTOM_RIGHT
                x < 1 / 3f -> TapZone.LEFT
                x > 2 / 3f -> TapZone.RIGHT
                else -> TapZone.CENTER
            }

            TWO_LEFT_RIGHT_CENTER -> when {
                x < 1 / 3f && y < 1 / 2f -> TapZone.TOP_LEFT
                x > 2 / 3f && y < 1 / 2f -> TapZone.TOP_RIGHT
                x < 1 / 3f && y >= 1 / 2f -> TapZone.BOTTOM_LEFT
                x > 2 / 3f && y >= 1 / 2f -> TapZone.BOTTOM_RIGHT
                y < 1 / 2f -> TapZone.TOP
                else -> TapZone.BOTTOM
            }

            THREE_LEFT_RIGHT_TWO_CENTER -> when {
                x < 1 / 3f && y < 1 / 3f -> TapZone.TOP_LEFT
                x > 2 / 3f && y < 1 / 3f -> TapZone.TOP_RIGHT
                x < 1 / 3f && y > 1 / 3f -> TapZone.BOTTOM_LEFT
                x > 2 / 3f && y > 1 / 3f -> TapZone.BOTTOM_RIGHT
                x < 1 / 3f -> TapZone.LEFT
                x > 2 / 3f -> TapZone.RIGHT
                y < 1 / 2f -> TapZone.TOP
                else -> TapZone.BOTTOM
            }

            TWO_LEFT_RIGHT_THREE_CENTER -> when {
                x < 1 / 3f && y < 1 / 2f -> TapZone.TOP_LEFT
                x > 2 / 3f && y < 1 / 2f -> TapZone.TOP_RIGHT
                x < 1 / 3f && y > 1 / 2f -> TapZone.BOTTOM_LEFT
                x > 2 / 3f && y > 1 / 2f -> TapZone.BOTTOM_RIGHT
                y < 1 / 3f -> TapZone.TOP
                y > 2 / 3f -> TapZone.BOTTOM
                else -> TapZone.CENTER
            }

            TRIANGLE_SIDE_CENTER_CORNERS -> {
                val inSquare = x >= 1 / 3f && x <= 2 / 3f && y >= 1 / 3f && y <= 2 / 3f
                val bottomOrLeft = y > x
                val topOrLeft = y < 1 - x

                when {
                    inSquare -> TapZone.CENTER
                    y < -x + 1 / 3f -> TapZone.TOP_LEFT
                    y < x - 2 / 3f -> TapZone.TOP_RIGHT
                    y > x + 2 / 3f -> TapZone.BOTTOM_LEFT
                    y > -x + 5 / 3f -> TapZone.BOTTOM_RIGHT
                    bottomOrLeft && topOrLeft -> TapZone.LEFT
                    !bottomOrLeft && topOrLeft -> TapZone.TOP
                    bottomOrLeft -> TapZone.RIGHT
                    else -> TapZone.BOTTOM
                }
            }

            TRIANGLE_SIDES -> {
                val bottomOrLeft = y > x
                val topOrLeft = y < 1 - x

                when {
                    bottomOrLeft && topOrLeft -> TapZone.LEFT
                    !bottomOrLeft && topOrLeft -> TapZone.TOP
                    bottomOrLeft -> TapZone.RIGHT
                    else -> TapZone.BOTTOM
                }
            }

            TRIANGLE_SIDES_CENTER -> {
                val inCenter = x >= 1 / 3f && x <= 2 / 3f && y >= 1 / 3f && y <= 2 / 3f
                val bottomOrLeft = y > x
                val topOrLeft = y < 1 - x

                when {
                    inCenter -> TapZone.CENTER
                    bottomOrLeft && topOrLeft -> TapZone.LEFT
                    !bottomOrLeft && topOrLeft -> TapZone.TOP
                    bottomOrLeft -> TapZone.RIGHT
                    else -> TapZone.BOTTOM
                }
            }

            TRIANGLE_SIDE_CORNERS -> {
                val bottomOrLeft = x < y
                val topOrLeft = x + y < 1

                when {
                    y < -x + 1 / 3f -> TapZone.TOP_LEFT
                    y < x - 2 / 3f -> TapZone.TOP_RIGHT
                    y > x + 2 / 3f -> TapZone.BOTTOM_LEFT
                    y > -x + 5 / 3f -> TapZone.BOTTOM_RIGHT
                    bottomOrLeft && topOrLeft -> TapZone.LEFT
                    !bottomOrLeft && topOrLeft -> TapZone.TOP
                    bottomOrLeft -> TapZone.RIGHT
                    else -> TapZone.BOTTOM
                }
            }
        }
    }
}